File: packages/runar-compiler/src/__tests__/constructor-args-validation.test.ts
/**
 * constructorArgs shape validation — `compile(opts.constructorArgs)` must
 * reject inputs that would silently bake nothing and emit placeholder
 * scripts that fail opaquely at runtime:
 *
 *  (a) positional arrays (natural guess, but keys match no property names)
 *  (b) keys that don't match any contract property (typos)
 *  (c) referenced readonly properties left unbaked after applying the args
 */

import { describe, it, expect } from 'vitest';
import { compile } from '../index.js';

// ---------------------------------------------------------------------------
// Contract sources
// ---------------------------------------------------------------------------

const HASH_LOCK_SOURCE = `
class HashLock extends SmartContract {
  readonly hashValue: Sha256;

  constructor(hashValue: Sha256) {
    super(hashValue);
    this.hashValue = hashValue;
  }

  public unlock(preimage: ByteString) {
    assert(sha256(preimage) === this.hashValue);
  }
}
`;

const TWO_PROP_SOURCE = `
class TwoProp extends SmartContract {
  readonly target: bigint;
  readonly unused: bigint;

  constructor(target: bigint, unused: bigint) {
    super(target, unused);
    this.target = target;
    this.unused = unused;
  }

  public check(x: bigint) {
    assert(x === this.target);
  }
}
`;

const HASH = 'aa'.repeat(32);

function errorMessages(result: ReturnType<typeof compile>): string {
  return result.diagnostics
    .filter((d) => d.severity === 'error')
    .map((d) => d.message)
    .join('\n');
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('constructorArgs validation', () => {
  it('rejects a positional array with a diagnostic error', () => {
    const result = compile(HASH_LOCK_SOURCE, {
      // Positional array — the shape RunarContract/TestSmartContract take,
      // but NOT what compile() takes. Previously baked nothing silently.
      constructorArgs: [HASH] as unknown as Record<string, string>,
    });

    expect(result.success).toBe(false);
    expect(errorMessages(result)).toMatch(/positional array/i);
  });

  it('rejects keys that match no contract property', () => {
    const result = compile(HASH_LOCK_SOURCE, {
      constructorArgs: { hashVal: HASH }, // typo: hashValue
    });

    expect(result.success).toBe(false);
    const msgs = errorMessages(result);
    expect(msgs).toContain("'hashVal'");
    expect(msgs).toContain('hashValue');
  });

  it('rejects when a referenced readonly property remains unbaked', () => {
    const result = compile(TWO_PROP_SOURCE, {
      // 'target' is referenced by check() but not provided.
      constructorArgs: { unused: 1n },
    });

    expect(result.success).toBe(false);
    expect(errorMessages(result)).toMatch(/'target'.*placeholder/s);
  });

  it('accepts an unreferenced readonly property left unbaked', () => {
    // 'unused' is never referenced by a method — DCE eliminates it, so
    // leaving it unbaked is fine.
    const result = compile(TWO_PROP_SOURCE, {
      constructorArgs: { target: 42n },
    });

    expect(result.success).toBe(true);
    expect(result.scriptHex).toBeDefined();
  });

  it('accepts a complete named record (baked script, no slots)', () => {
    const result = compile(HASH_LOCK_SOURCE, {
      constructorArgs: { hashValue: HASH },
    });

    expect(result.success).toBe(true);
    expect(result.artifact!.script).toContain(HASH);
    const slots = result.artifact!.constructorSlots;
    expect(slots === undefined || slots.length === 0).toBe(true);
  });

  it('still compiles placeholder artifacts when no constructorArgs given', () => {
    const result = compile(HASH_LOCK_SOURCE);
    expect(result.success).toBe(true);
    expect(result.artifact!.constructorSlots!.length).toBeGreaterThanOrEqual(1);
  });
});

File: compilers/go/compiler/constructor_args_validation_test.go
package compiler

// Port of the TypeScript reference test
// packages/runar-compiler/src/__tests__/constructor-args-validation.test.ts.
//
// applyConstructorArgs must reject inputs that would silently bake nothing and
// emit placeholder scripts that fail opaquely at runtime:
//
//	(a) positional arrays — N/A for the Go tier: ConstructorArgs is a typed
//	    map[string]interface{}, so a positional array cannot reach the API.
//	    (The dynamically-typed tiers guard this explicitly.)
//	(b) keys that match no contract property (typos)
//	(c) referenced readonly properties left unbaked after applying the args
//
// The placeholder path (no ConstructorArgs) stays unchecked / byte-identical.

import (
	"math/big"
	"strings"
	"testing"

	"github.com/icellan/runar/compilers/go/frontend"
)

const hashLockSource = `import { SmartContract, assert, sha256 } from 'runar-lang';
import type { ByteString, Sha256 } from 'runar-lang';

class HashLock extends SmartContract {
  readonly hashValue: Sha256;

  constructor(hashValue: Sha256) {
    super(hashValue);
    this.hashValue = hashValue;
  }

  public unlock(preimage: ByteString) {
    assert(sha256(preimage) === this.hashValue);
  }
}
`

const twoPropSource = `import { SmartContract, assert } from 'runar-lang';

class TwoProp extends SmartContract {
  readonly target: bigint;
  readonly unused: bigint;

  constructor(target: bigint, unused: bigint) {
    super(target, unused);
    this.target = target;
    this.unused = unused;
  }

  public check(x: bigint) {
    assert(x === this.target);
  }
}
`

const hashHex = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" // "aa" * 32

func errorMessages(result *CompileResult) string {
	var msgs []string
	for _, d := range result.Diagnostics {
		if d.Severity == frontend.SeverityError {
			msgs = append(msgs, d.Message)
		}
	}
	return strings.Join(msgs, "\n")
}

func TestConstructorArgs_RejectsUnknownKey(t *testing.T) {
	result := CompileFromSourceStrWithResult(hashLockSource, "HashLock.runar.ts", CompileOptions{
		ConstructorArgs: map[string]interface{}{"hashVal": hashHex}, // typo: hashValue
	})

	if result.Success {
		t.Fatalf("expected compilation to fail for an unknown constructorArgs key")
	}
	msgs := errorMessages(result)
	if !strings.Contains(msgs, "'hashVal'") {
		t.Errorf("error should name the offending key 'hashVal'; got:\n%s", msgs)
	}
	if !strings.Contains(msgs, "hashValue") {
		t.Errorf("error should list the valid property 'hashValue'; got:\n%s", msgs)
	}
}

func TestConstructorArgs_RejectsUnbakedReferencedReadonly(t *testing.T) {
	// 'target' is referenced by check() but not provided; only 'unused' is baked.
	result := CompileFromSourceStrWithResult(twoPropSource, "TwoProp.runar.ts", CompileOptions{
		ConstructorArgs: map[string]interface{}{"unused": big.NewInt(1)},
	})

	if result.Success {
		t.Fatalf("expected compilation to fail when a referenced readonly stays unbaked")
	}
	msgs := errorMessages(result)
	if !strings.Contains(msgs, "'target'") {
		t.Errorf("error should name the unbaked referenced property 'target'; got:\n%s", msgs)
	}
	if !strings.Contains(msgs, "placeholder") {
		t.Errorf("error should mention the OP_0 placeholder hazard; got:\n%s", msgs)
	}
}

func TestConstructorArgs_AcceptsUnreferencedUnbakedReadonly(t *testing.T) {
	// 'unused' is never referenced by a method — DCE eliminates it, so leaving
	// it unbaked is fine.
	result := CompileFromSourceStrWithResult(twoPropSource, "TwoProp.runar.ts", CompileOptions{
		ConstructorArgs: map[string]interface{}{"target": big.NewInt(42)},
	})

	if !result.Success {
		t.Fatalf("expected success (unused readonly is unreferenced); diagnostics:\n%s", errorMessages(result))
	}
	if result.Artifact == nil || result.Artifact.Script == "" {
		t.Errorf("expected a compiled artifact with a script")
	}
}

func TestConstructorArgs_AcceptsCompleteRecord(t *testing.T) {
	result := CompileFromSourceStrWithResult(hashLockSource, "HashLock.runar.ts", CompileOptions{
		ConstructorArgs: map[string]interface{}{"hashValue": hashHex},
	})

	if !result.Success {
		t.Fatalf("expected success for a complete named record; diagnostics:\n%s", errorMessages(result))
	}
	if result.Artifact == nil {
		t.Fatal("expected a compiled artifact")
	}
	if !strings.Contains(result.Artifact.Script, hashHex) {
		t.Errorf("baked script should contain the hash bytes %s; got script:\n%s", hashHex, result.Artifact.Script)
	}
	if len(result.Artifact.ConstructorSlots) != 0 {
		t.Errorf("a fully baked contract should carry no constructor slots; got %d", len(result.Artifact.ConstructorSlots))
	}
}

func TestConstructorArgs_NoArgsStillProducesPlaceholderSlots(t *testing.T) {
	// The no-args placeholder path stays unchecked and byte-identical to before.
	result := CompileFromSourceStrWithResult(hashLockSource, "HashLock.runar.ts", CompileOptions{})

	if !result.Success {
		t.Fatalf("expected success for placeholder compilation; diagnostics:\n%s", errorMessages(result))
	}
	if result.Artifact == nil || len(result.Artifact.ConstructorSlots) < 1 {
		t.Errorf("placeholder compilation should carry at least one constructor slot")
	}
}
